### ISSUE TITLE
C# Google.Cloud.Spanner.Data Driver - Query 50K records - throw NPE

### ISSUE THREAD

**Issue by alyaros** [googleapis/google-cloud-dotnet#2777]
Version:
Google.Cloud.Spanner.Data -Version 2.0.0-beta05.
Google.Cloud.Spanner.V1 -Version 2.0.0-beta05

**Simple select query "select * from My50KRecordsTable" throw NPE.**

Object reference not set to an instance of an object.

   at Google.Cloud.Spanner.V1.ReliableStreamReader.MergeChunk(Value currentValue, Value nextValue)
   at Google.Cloud.Spanner.V1.ReliableStreamReader.<NextAsync>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.SpannerDataReader.<>c__DisplayClass57_0.<<ReadAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Google.Cloud.Spanner.Data.ExecuteHelper.<WithErrorTranslationAndProfiling>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Google.Api.Gax.TaskExtensions.WaitWithUnwrappedExceptions(Task task)
   at Google.Cloud.Spanner.Data.SpannerDataReader.Read()
   at NG.EntityFramework.Spanner.Storage.Read.DbDataReaderDecorator.Read()
   at Microsoft.EntityFrameworkCore.Storage.RelationalDataReader.Read()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.Storage.Internal.NoopExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.<_TrackEntities>d__17`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()

**Comment by jskeet** [comment]
Eek, that's annoying/worrying/irksome.

I'm on vacation now until January, but hopefully Amanda will be able to look into it. Could you tell us what the structure of your table is, just in terms of the schema? That couls easily affect things.

**Comment by alyaros** [comment]
Something random for testing:

UniqueId | STRING(MAX) | Yes
CreationDate | TIMESTAMP | Yes
IsRandom | BOOL | Yes
RandomNumber | INT64 | Yes

I've just made few more attempts with 60K and it worked so this is not stable in terms for reproducing it.

**Comment by amanda-tarafa** [comment]
I'll take a look at this this week. Will aim for today else on Friday.

**Comment by amanda-tarafa** [comment]
@alyaros I haven't been able to reproduce this.

I've created a table with the same schema that you provided. I filled that table using the following code:
```
private static async Task FillTable()
{
    string connectionString = "data source=projects/<project_id>/instances/<instance_id>/databases/<database_id>";

    using (SpannerConnection connection = new SpannerConnection(connectionString))
    {
        await connection.OpenAsync();

        using (SpannerTransaction transaction = await connection.BeginTransactionAsync())
        {
            transaction.CommitTimeout = 0;
            string insert = "INSERT INTO <table_id> (UniqueID, CreationDate, IsRandom, RandomNumber) VALUES (@UniqueID, @CreationDate, @IsRandom, @RandomNumber)";
            using (var command = connection.CreateDmlCommand(insert))
            {
                command.Transaction = transaction;
                Random random = new Random();
                DateTime now = DateTime.UtcNow;

                for(int i = 0; i <= 2_000; i++)
                {
                    command.Parameters.Add("UniqueID", SpannerDbType.String, Guid.NewGuid().ToString());
                    command.Parameters.Add("CreationDate", SpannerDbType.Timestamp, now.AddDays(random.Next(-10_000, 0)));
                    command.Parameters.Add("IsRandom", SpannerDbType.Bool, random.Next(2) == 0 ? false : true);
                    command.Parameters.Add("RandomNumber", SpannerDbType.Int64, random.Next(int.MinValue, int.MaxValue));
                    if (random.Next(1000) == 0)
                    {
                        switch(random.Next(4))
                        {
                            case 0:
                                command.Parameters["CreationDate"].Value = null;
                                break;
                            case 1:
                                command.Parameters["IsRandom"].Value = null;
                                break;
                            case 2:
                                command.Parameters["RandomNumber"].Value = null;
                                break;
                            default:
                                break;
                        }
                    }
                    await command.ExecuteNonQueryAsync();
                    command.Parameters.Clear();
                }
                await transaction.CommitAsync();
            }
        }
    }
}
```
I ran the above code several times, querying the data in between, until I got to 50k rows. As you can see I've added some null values here and there in case that was the source of the problem. Also I manually added a couple other rows, one of them with all null values.

Then I've been running the following code for a few hours without it failing:
```
private static async Task RunQuery()
{
    string connectionString = "data source=projects/<project_id>/instances/<instance_id>/databases/<database_id>";

    do
    {
        using (SpannerConnection connection = new SpannerConnection(connectionString))
        {
            await connection.OpenAsync();

            SpannerCommand selectCmd = connection.CreateSelectCommand("SELECT * FROM <table_id>");
            using (SpannerDataReader reader = await selectCmd.ExecuteReaderAsync())
            {
                while (reader.Read())
                {
                    var id = reader["UniqueID"];
                    var date = reader["CreationDate"];
                    var isRandom = reader["IsRandom"];
                    var number = reader["RandomNumber"];
                }
            }
        }

        Thread.Sleep(10_000);
    }
    while (true);
}
```

So a few asks for you to see if we can reproduce this and attempt a fix:

- Can you confirm that your data is not essentially different than ours? Are your strings a similar length to ours etc?
- Can you run the method `RunQuery` against your 50K row table and see if it fails? Just change the connection string and the name of the table.
- I've seen from the stack trace that there's some custom EntityFramework wrapper (`NG.EntityFramework.Spanner.Storage.Read.DbDataReaderDecorator`) on top of our `SpannerDataReader`, so, especially if `RunQuery` works on your table but your original code still doesn't could you provide us with a stripped down version of your code (as similar to `RunQuery` as possible) that does fails on your 50K table?

Thanks!

**Comment by alyaros** [comment]
Hey,

The data is the same, nothing special in my case.

Although EF is involved and maybe has some contribution to the case - i.e in some strange way it generates it - I think it is less relevant. You can think of it as just another caller\user of your code - as bad as it can be :)

Anyway, Unfortunately, the issue isn't reproducible in my code as well - originally, I ran it for several times and got the 50K NPE and afterwards (also now) it worked without a problem - So i can now only offer the error itself.

Maybe it's worth reviewing it again and\or write it in a more defensive way so when it will get thrown again somewhere you'll have more details.

BTW, Thanks a lot for the quick response :)

**Comment by amanda-tarafa** [comment]
As soon as I'll get back from holidays I'll take a closer look at the code to see if I can find where and why the exception might be thrown and then do something about it. I'll update here with whatever I find.

**Comment by jskeet** [comment]
I'm having a look at this code now at the same time that I'm looking at #2779.

Just to confirm, your table doesn't have any non-scalar types in (structs or lists)?
Could you tell us the rough size of the strings in the table that you *did* see the problem in?

(It looks like MergeChunks may have some slightly questionable code that we should fix, but that's mostly related to structs and lists, not scalars.)

**Comment by jskeet** [comment]
That's odd - that's exactly the type of ID I've created in #2779.

Just to check, do you have multiple threads using the same reader anywhere? That could potentially explain both the issues you're seeing.

**Comment by jskeet** [comment]
Just as a heads-up, I've now managed to reproduce #2779 against @alyaros's database, but not this issue yet. I'm still looking though.

**Comment by alyaros** [comment]
I think that it's worth trying running those queries in a small stress mode - just write several related queries over this table and run them in the same process one after another for some time. Eventually you'll get NPE. (From my experience this happened quite quickly - after few minutes of run)

**Comment by jskeet** [comment]
That's good to know - thanks!

**Comment by jskeet** [comment]
I still haven't seen the NRE, but I've seen an InvalidOperationException due to the response not including as many values as expected (i.e. not a multiple of the number of fields). If that happened in the context of a merge, that could easily provoke the NRE. So my working assumption is that that's the underlying cause, and I'll try to work out what's going on there. (It may be to do with our exception recovery, where we resume from a certain point and skip a number of values.)

**Comment by jskeet** [comment]
Closing as this is fixed in 2.0.0-beta06.
